src/game/common/rts/team.cpp

```diff
@@ -0,0 +1,815 @@
+    xfer->xferUnsignedShort(&size);
+
+    if (xfer->Get_Mode() == XFER_SAVE) {
+        for (auto it = m_relationships.begin(); it != m_relationships.end(); ++it) {
+            unsigned int id = (*it).first;
+            xfer->xferUser(&id, sizeof(id));
+            Relationship r = (*it).second;
+            xfer->xferUser(&r, sizeof(r));
+        }
+    } else {
+        for (int i = 0; i < size; i++) {
+            unsigned int id;
+            xfer->xferUser(&id, sizeof(id));
+            Relationship r;
+            xfer->xferUser(&r, sizeof(r));
+            m_relationships[i] = r;
+        }
+    }
+}
+
+TeamFactory::TeamFactory() : m_nextPrototypeID(0), m_nextTeamID(0) {}
+
+TeamFactory::~TeamFactory()
+{
+    Clear();
+}
+
+void TeamFactory::Init()
+{
+    Clear();
+}
+
+void TeamFactory::Reset()
+{
+    m_nextPrototypeID = 0;
+    m_nextTeamID = 0;
+    Clear();
+}
+
+void TeamFactory::Xfer_Snapshot(Xfer *xfer)
+{
+    unsigned char version = 1;
+    xfer->xferVersion(&version, 1);
+    xfer->xferUser(&m_nextTeamID, sizeof(m_nextTeamID));
+    unsigned short size = (unsigned short)m_prototypes.size();
+    xfer->xferUnsignedShort(&size);
+    captainslog_relassert(size == m_prototypes.size(),
+        6,
+        "TeamFactory::xfer - Prototype count mismatch '%d should be '%d'",
+        size,
+        m_prototypes.size());
+    std::map<NameKeyType, TeamPrototype *>::iterator it;
+    Utf8String str;
+
+    if (xfer->Get_Mode() == XFER_SAVE) {
+        for (it = m_prototypes.begin(); it != m_prototypes.end(); ++it) {
+            TeamPrototype *proto = it->second;
+            unsigned int id = proto->Get_ID();
+            xfer->xferUser(&id, sizeof(id));
+            xfer->xferSnapshot(proto);
+        }
+    } else {
+        for (int i = 0; i < size; i++) {
+            unsigned int id;
+            xfer->xferUser(&id, sizeof(id));
+            TeamPrototype *proto = Find_Team_Prototype_By_ID(id);
+            captainslog_relassert(proto != nullptr, 6, "TeamFactory::xfer - Unable to find team prototype by id");
+            xfer->xferSnapshot(proto);
+        }
+    }
+}
+
+void TeamFactory::Load_Post_Process()
+{
+    m_nextTeamID = 0;
+    m_nextPrototypeID = 0;
+    std::map<NameKeyType, TeamPrototype *>::iterator it;
+
+    for (it = m_prototypes.begin(); it != m_prototypes.end(); ++it) {
+        TeamPrototype *proto = it->second;
+
+        if (proto->Get_ID() >= m_nextPrototypeID) {
+            m_nextPrototypeID = proto->Get_ID() + 1;
+        }
+
+        DLINK_ITERATOR<Team> iter = proto->Iterate_Team_Instance_List();
+
+        while (!iter.Done()) {
+            Team *team = iter.Cur();
+
+            if (team->Get_Team_ID() >= m_nextTeamID) {
+                m_nextTeamID = team->Get_Team_ID() + 1;
+            }
+
+            iter.Advance();
+        }
+    }
+}
+
+void TeamFactory::Clear()
+{
+    std::map<NameKeyType, TeamPrototype *> temp(m_prototypes);
+    m_prototypes.clear();
```

Comment:
This does not look efficient. Better:

```
std::map<NameKeyType, TeamPrototype *> temp;
temp.swap(m_prototypes);
```

Result is same, but avoids copy of whole map.

Reply:
This is what the original does and I would prefer to keep it original.